File: 67.二进制求和.py
#
# @lc app=leetcode.cn id=67 lang=python3
#
# [67] 二进制求和
#

# @lc code=start

# @lc code=end
print(Solution().addBinary("111","1010"))

# solution1:
# class Solution:
#     def addBinary(self, a: str, b: str) -> str:
#         num1=int(a,2)
#         num2=int(b,2)
#         sum=bin(num1+num2)
#         return str(sum)[2:]
# class Solution:
#     def addBinary(self, a, b) -> str:
#         return '{:b}'.format(int(a, 2) + int(b, 2))

# Solution2:模拟
# class Solution:
#     def addBinary(self, a: str, b: str) -> str:
#         ans=''
#         carry,lenth1,lenth2=0,len(a),len(b)
#         if lenth1<=lenth2:
#             a=(lenth2-lenth1)*'0'+a
#             lenth1=lenth2
#         else:
#             b=(lenth1-lenth2)*'0'+b
#             lenth2=lenth1
#         for i in range(lenth1-1,-1,-1):
#             # 法一
#             tmp=int(a[i])^int(b[i])^carry
#             ans+=str(tmp)
#             # if (int(a[i])+int(b[i])+carry)>=2:
#             #     carry=1
#             # else:
#             #     carry=0
#             carry=(int(a[i])+int(b[i])+carry)//2
#             #法二
#             # tmp=int(a[i])+int(b[i])+carry
#             # ans+=str(tmp%2)
#             # carry=tmp//2
#         if carry==1:
#             ans+=str(carry)
#         return ans[::-1]


# ?solution3:
# class Solution:
#     def addBinary(self, a, b) -> str:
#         x, y = int(a, 2), int(b, 2)
#         while y:
#             answer = x ^ y
#             carry = (x & y) << 1
#             x, y = answer, carry
#         return bin(x)[2:]

File: 67.二进制求和.java
/*
 * @lc app=leetcode.cn id=67 lang=java
 *
 * [67] 二进制求和
 */

// @lc code=start
class Solution {
    public String addBinary(String a, String b) {
        int m=a.length()-1,n=b.length()-1;
        StringBuilder stringBuilder=new StringBuilder();
        int carry=0;
        while(m>=0&&n>=0){
            int sum=carry;
            sum+=a.charAt(m--)-'0';
            sum+=b.charAt(n--)-'0';
            carry=sum/2;
            stringBuilder.append(sum%2);
        }
        while(m>=0){
            int sum=carry+a.charAt(m--)-'0';
            carry=sum/2;
            stringBuilder.append(sum%2);
        }
        while(n>=0){
            int sum=carry+b.charAt(n--)-'0';
            carry=sum/2;
            stringBuilder.append(sum%2);
        }
        if(carry==1){
            stringBuilder.append(1);
        }
        return stringBuilder.reverse().toString();
    }
}
// @lc code=end

//严重错误 溢出
// Solution1:
// class Solution {
//     public String addBinary(String a, String b) {
//         return  Integer.toString((Integer.valueOf(a,2)+Integer.valueOf(b,2)),2);
//     }
// }

// 补全两个字符串至一样长
// Solution2：
// class Solution {
//     public String addBinary(String a, String b) {
//         int m=a.length(),n=b.length();
//         int carry=0;
//         StringBuilder stringBuilder=new StringBuilder();
//         if(m<n){
//             for(int i=0;i<n-m;i++){
//                 stringBuilder.append(0);
//             }
//             a=stringBuilder.toString()+a;
//         }else{
//             for(int i=0;i<m-n;i++){
//                 stringBuilder.append(0);
//             }
//             b=stringBuilder.toString()+b;
//         }
//         StringBuilder stringBuilder2=new StringBuilder();
//         for(int i=Math.max(m, n)-1;i>=0;i--){
//            int sum=a.charAt(i)-'0'+b.charAt(i)-'0'+carry;
//             stringBuilder2.append(sum%2);
//             carry=sum/2;
//         }
//         if(carry==1){
//             stringBuilder2.append(1);
//         }
//         return stringBuilder2.reverse().toString();
//     }
// }





// Solution3:
// class Solution {
//     public String addBinary(String a, String b) {
//         int m=a.length()-1,n=b.length()-1;
//         StringBuilder stringBuilder=new StringBuilder();
//         int carry=0;
//         while(m>=0&&n>=0){
//             int sum=carry;
//             sum+=a.charAt(m--)-'0';
//             sum+=b.charAt(n--)-'0';
//             carry=sum/2;
//             stringBuilder.append(sum%2);
//         }
//         while(m>=0){
//             int sum=carry+a.charAt(m--)-'0';
//             carry=sum/2;
//             stringBuilder.append(sum%2);
//         }
//         while(n>=0){
//             int sum=carry+b.charAt(n--)-'0';
//             carry=sum/2;
//             stringBuilder.append(sum%2);
//         }
//         if(carry==1){
//             stringBuilder.append(1);
//         }
//         return stringBuilder.reverse().toString();
//     }
// }